### ISSUE TITLE
Mixins can lead to empty exposure sets in weird ways

### ISSUE THREAD

**Issue by bzbarsky** [heycam/webidl#761]
Consider this IDL:
```
[Exposed=Window] interface Foo {};
[Exposed=Worker] interface mixin Bar { 
  readonly attribute long attr; 
};
Foo includes Bar;
```
If I read https://heycam.github.io/webidl/#Exposed correctly, this should result in `attr` having an empty exposure set on `Foo`.  In fact, the entire `includes` statement is pointless: given the exposure set of the mixin, it can't add anything to the interface.  This would be a little less obvious if it were written like so:
```
[Exposed=Window] interface Foo {};
interface mixin Bar { 
  [Exposed=Worker] 
  readonly attribute long attr; 
};
Foo includes Bar;
```
because then adding an un-annotated member to the mixin would cause it to be exposed on the interface, so that part would be OK, in theory.

If we're generally OK with this setup, we should still consider making that first snippet invalid, so people don't accidentally write it.

As a generalization of that, I'd actually prefer it if mixins had to explicitly say which globals they're OK being mixed into and if we required that the exposure set of a mixin had to be a superset of the exposure sets of the interfaces it's mixed into.  Members that shouldn't be exposed in all those would then have their own Exposed attributes or be in a `partial interface mixin` with an `Exposed` attribute...  I guess that's basically the behavior we get right now if the mixin does not specify `Exposed` at all, but it would be a bit more explicit...

@saschanaz @Ms2ger @tobie @domenic @annevk

**Comment by annevk** [comment]
Why a superset? If I have interface X exposed to A and B, it seems reasonable to include a mixin Y into X that is exposed to A or B. It does seem reasonable to me that if there's no overlap the inclusion should be invalid.

(Aside: it's always a little weird to me that we have this "extended attribute" concept that's sometimes really core to the infrastructure and sometimes rather optional when evaluating some IDL, but I guess that's how it is and changing that is a large-scale operation best postponed until we have something really good.)

**Comment by bzbarsky** [comment]
Right, that is the argument for the spec's current intersection setup.  My main concern with it is the somewhat implicit (not explicitly stated in the IDL) way in which the exposure set is determined.  In particular, if I have:
```
[Exposed=(A,B)]
interface Foo {};

[Exposed=(B,C)]
interface mixin Bar {};

Foo includes Bar;
```
then reading the definition of a mixin does not make it clear that some things from the mixin will only get exposed in B and not any other global.   But I guess things would be no better if the mixin overall had `Exposed=(A,B,C)` and then a specific member had `Exposed=(B,C)`....

**Comment by annevk** [comment]
I think as long as you see a top-level [Exposed] declaration as a shorthand for a (default) per-member declaration it's reasonably clear where members end up.
